docs/src/developer/design/user_api/add_target.rst

```diff
@@ -0,0 +1,309 @@
+
+.. glossary::
+
+   executable
+      A program meant to be run by a user. It may be compiled or it may not be,
+      e.g., a Python script with a ``main`` function.
+
+   library
+      A collection of functionality distributed as a single packaged entity.
+      Libraries are designed for consumption by other packages.
+
+   source file
+      Any file containing code. For languages like C++, "source file" includes
+      header files.
+
+As a slight aside, we choose these terms in order to conform to CMake's already
+existing :term:`API`.
+
+**********************************
+Add Target Function Considerations
+**********************************
+
+.. _at_recording_targets:
+
+recording targets
+   The primary motivation for the add target functions is to serve as a
+   mechanism for recording the details of the target.
+
+.. _at_succinctness:
+
+succinctness
+   A lot of the information CMake's add target functions require can be
+   gleaned from other sources. Requiring the user to restate the information is
+   verbose and violates :term:`DRY`.
+
+.. _at_coding_languages:
+
+coding language
+   Exactly what targets can be built/found depend on the coding language(s) of
+   the project. For example, shared/static :term:`libraries <library>` do not
+   exist in the context of building a Python code, but do exist when building
+   C/C++ code.
+
+   - Targeted coding languages include: C, C++, CMake, Fortran, Python, and
+     extensions of the aforementioned languages (e.g., CMaize, CUDA and OpenMP).
+
+.. _at_target_sources:
+
+target sources
+   Targets are usually associated with :term:`source files<source file>`.
+
+   - As a corollary we note that source files usually fall into two categories,
+     public and private. Public source files need to be redistributed with the
+     target, whereas private source files are consumed in building the target.
+
+.. _at_conditional_targets:
+
+conditional targets
+   Many projects contain targets which are only conditionally built. These
+   targets may be optional package features, or targets only needed for testing
+   or maintaining the project. In our experience, in a traditional CMake-based
+   :term:`build system`, it is rarely possible to isolate the logic for these
+   targets because CMake requires them to be specified in multiple places. We
+   assume the following about a conditional target:
+
+   - is conditionally included based on the value of a variable (if it's
+     actually multiple variables, the user, via boolean logic, can combine the
+     variables into a single variable). For optional features the variable is
+     usually something like ``ENABLE_XXX``; for tests the variable is
+     ``BUILD_TESTING`` (defined by CMake).
+   - needs to be conditionally built, linked to, tested, and packaged. In other
+     words, when CMaize is given a list of targets, CMaize needs to skip
+     conditional targets which are not currently enabled.
+
```

Comment:
I would suggest adding another consideration here, something like "maintaining best practices" saying:

Along with the concerns addressed in consideration `succinctness`, creating a new target in CMake has different best practices for different languages and different target systems that often must be inferred or learned from other sources.

Reply:
I like this idea, though I'm not sold on your wording (or the wording I just pushed). Let me know what you think.

Reply:
I think your new wording is fine!